braille-printer-app/brf-printer-app.c

```diff
@@ -0,0 +1,821 @@
+  filter_data->back_pipe[0] = -1;
+
+  filter_data->back_pipe[1] = -1;
+  filter_data->side_pipe[0] = -1;
+  filter_data->side_pipe[1] = -1;
+  filter_data->logdata = job;
+  //filter_data->iscanceledfunc = papplJobIsCanceled(job); // Function to indicate
+                                                         // whether the job got
+                                                         // canceled
+  filter_data->iscanceleddata = job;
+  // //
+  // // Load the printer's assigned PPD file, and find out which PPD option
+  // // seetings correspond to our job options
+  // //
+
+  // papplLogJob(job, PAPPL_LOGLEVEL_DEBUG,
+  //             "Printing job in spooling mode");
+
+  // filter_data_ext =
+  //     (ppd_filter_data_ext_t *)cf_filter_external_tilterDataGetExt(filter_data,
+  //                                                 PPD_FILTER_DATA_EXT);
+
+  //
+  // Open the input file...
+  //
+
+  filename = papplJobGetFilename(job);
+  if ((fd = open(filename, O_RDONLY)) < 0)
+  {
+    papplLogJob(job, PAPPL_LOGLEVEL_ERROR, "Unable to open input file '%s' for printing: %s",
+                filename, strerror(errno));
+    return (false);
+  }
+
+  //
+  // Get input file format
+  //
+
+  informat = papplJobGetFormat(job);
+  papplLogJob(job, PAPPL_LOGLEVEL_DEBUG,
+              "Input file format: %s", informat);
+  //
+  // Passing values to ppdFilterExternalCUPS()
+  //
+  filter_data_ext = (cf_filter_external_t *)calloc(1, sizeof(cf_filter_external_t));
+  
+  filter_data_ext->filter = "usr/lib/cups/filter/txttobrf"; 
+  filter_data_ext->num_options =0;
+  filter_data_ext->options = NULL;
+  filter_data_ext->envp= NULL;
+
+
+  //
+  // Find filters to use for this job
+  //
+
+  for (conversion =
+           (brf_spooling_conversion_t *)
+               cupsArrayFirst(spooling_conversions);
+       conversion;
+       conversion =
+           (brf_spooling_conversion_t *)
+               cupsArrayNext(spooling_conversions))
+   
+  {
+    if (strcmp(conversion->srctype, informat) == 0)
+
+      break;
+    
+  }
+  if (conversion == NULL )
+  {
+    papplLogJob(job, PAPPL_LOGLEVEL_ERROR,
+                "No pre-filter found for input format %s",
+                informat);
+    return (false);
+  }
+  // Set input and output formats for the filter chain
+  filter_data->content_type = conversion->srctype;
+  filter_data->final_content_type = conversion->dsttype;
+
+  //
+  // Connect the job's filter_data to the backend
+  //
+
+  if (strncmp(device_uri, "cups:", 5) == 0)
+  {
+    // Get the device data
+    device_data = (brf_cups_device_data_t *)papplDeviceGetData(device);
+
+    // Connect the filter_data
+    device_data->filter_data = filter_data;
+  }
+
+  //
+  // Set up filter function chain
+  //
+
+  chain = cupsArrayNew(NULL, NULL);
+
+  for (int i = 0; i < conversion->num_filters; i++){
+    cupsArrayAdd(chain, &(conversion->filters[i]));
+    cupsArrayAdd(chain,&(filter_data_ext));
+    }
+    
```

Comment:
The filter chain `chain` contains conversion filters from `conversion`. These must be of type `cf_filter_filter_in_chain_t`, but `filter_data_ext` is of type `cf_filter_external_t`. Also does it have to be added to `chain` after each of the conversion filters?